borg/remote.py

```diff
@@ -28,6 +29,21 @@
 MAX_INFLIGHT = 100
 
 
+def os_write(fd, data):
+    """os.write wrapper so we do not lose data for partial writes."""
+    # This is happening frequently on cygwin due to its small pipe buffer size of only 64kiB
+    # and also due to its different blocking pipe behaviour compared to Linux/*BSD.
+    # Neither Linux nor *BSD ever do partial writes on blocking pipes, unless interrupted by a
+    # signal, in which case serve() would terminate.
+    amount = len(data)
+    while data:
+        count = os.write(fd, data)
+        data = data[count:]
```

Comment:
> Using count == 0 as termination criteria: i don't think this is the correct criteria. Also, even if it would never return 0 if there was something to send, it would need an extra, empty write call to return 0.

Yes it'd have to be "not amount"; updating amount. to_send uses a somewhat similar construction, it also does copies. Might be an avenue for some minor optimization.

> I added the sleep to avoid too fast looping there. If it does a partial write, it means the buffer is currently full, so it might be rather pointless to just try again immediately without waiting a bit and give it a chance to empty the buffer.

Given that cygwin is quite strange here already this might be reasonable; it *does* interpret "blocking" quite different from all other OSes as you have found, so it might just become non-blocking when the pipe's buffer is full.

(Which I think would be 1-2 cygwin bugs).